### PR TITLE
GPU CUDA delayed updates

### PR DESCRIPTION
**Pull request by atillack**
To get things moving forward here is the the pull request for my GPU delayed updates code. I've been presenting some aspects of this work (including profiles) at the annual ECP meeting and at Nvidia's GTC earlier this year (2018) for those interested to get a more in depth view. It is labeled work in progress and here is my todo-list:

- [x] implement VMC, VMC w/ drift, and DMC parts

- [x] a bit of code merging is needed to use the config changes from Ye's CPU delayed updates instead of the ones I've been using
  - only difference is configurability per QMC block instead of global in the slaterdeterminant definition but I can live with losing that

- [x] extend to complex code path

- [x] code cleanup (leftover from some of the different strategies tried)

I tested and profiled the code extensively using the NiO system and see about a 1.5x speedup for DMC blocks of the 256 atom NiO on Summit. This is about what one would expect if the runtime of the update_inverse kernels were reduced to close to nothing per update step.

### CODE REVIEW

**Comment by PDoakORNL** [comment]
ok to test

**Comment by prckent** [comment]
To help development & merge velocity, you could do the complex implementation in a later PR if you wished. The GPU code was not complex until #84 by you and @yingwaili

**Comment by ye-luo** [comment]
Okay to test

File: src/Numerics/CUDA/cuda_inverse.h [hunk 1]
```diff
@@ -29,6 +29,26 @@ cublas_inverse (cublasHandle_t handle,
                 int N, int rowStride, int numMats,
                 bool useHigherPrecision = true);
 
+void
+cublas_lemma_mats (cublasHandle_t handle,
```
Comment (ye-luo) [inline src/Numerics/CUDA/cuda_inverse.h]: cuda_inverse.h/cu is for matrix inversion.
cublas_lemma_mats, cublas_ainv_row, cublas_smw_update are not generic wrapper functions of cublas.
Please move them to /sandbox/opt/cleanup/atillack/src/QMCWaveFunctions/Fermion

**Comment by ye-luo** [comment]
SM1 needs a fix at the moment. Regarding the complex, you can either fix the complex or protect the complex with macro depending on how much effort is needed.

File: src/Numerics/CUDA/cuda_inverse.h (hunk 1 above)
Comment (prckent) [reply src/Numerics/CUDA/cuda_inverse.h]: No production code should be in sandbox. Put distinct functionality in different files or simply rename the cuda_inverse.h file e.g. cuda_matrices.h

Comment (ye-luo) [reply src/Numerics/CUDA/cuda_inverse.h]: The path I put previously was wrong. It was just my local path. Functions under Numerics should not be only used a specific algorithm. So cuda_matrices.h is not good. Please create a new set of .h and .cu file under src/QMCWaveFunctions/Fermion.

Comment (prckent) [reply src/Numerics/CUDA/cuda_inverse.h]: OK

**Comment by atillack** [comment]
Happy New Year! The hamiltonian unit test should work now. Thank you @ye-luo for your fix.

**Comment by atillack** [comment]
Complex code path implemented and tested to be working on NiO S32 on SummitDev.

File: src/Numerics/CUDA/cuda_inverse.h (hunk 1 above)
Comment (atillack) [reply src/Numerics/CUDA/cuda_inverse.h]: @ye-luo @prckent Done.

**Comment by prckent** [comment]
Is short-diamondC_2x1x1_pp-vmc_sdj-1-16 still failing?

**Comment by atillack** [comment]
@ye-luo It is working now - thanks for the good test case :-)

**Comment by ye-luo** [comment]
There is still bug in the code.
short-NiO_a4_e48_pp-delayed_update-dmc_sdj-1-16 is failing.
This test is not turned on by default. You will need to provide the NiO S1 h5 via QMC_DATA like the performance test.

I also ran NiO 128 atom cell on Summit and the first vmc section went wrong. The variance needs to be all around 420.
```
$ qmca -q ev dmc*/*.s001.scalar.dat
                            LocalEnergy               Variance           ratio 
dmc-a128-e1536-cpu-SM/NiO-fcc-S32-dmc  series 1  -11863.571219 +/- 0.625023   427.132270 +/- 18.436181   0.0360 
dmc-a128-e1536-cpu-DU02/NiO-fcc-S32-dmc  series 1  -11864.226169 +/- 1.851689   454.180177 +/- 59.596102   0.0383 
dmc-a128-e1536-cpu-DU04/NiO-fcc-S32-dmc  series 1  -11865.875291 +/- 0.208591   472.741415 +/- 45.019062   0.0398 
dmc-a128-e1536-cpu-DU08/NiO-fcc-S32-dmc  series 1  -11870.404495 +/- 4.905789   835.706557 +/- 306.005369   0.0704 
dmc-a128-e1536-cpu-DU12/NiO-fcc-S32-dmc  series 1  -11878.254077 +/- 10.213142   1463.242556 +/- 729.916064   0.1232 
dmc-a128-e1536-cpu-DU16/NiO-fcc-S32-dmc  series 1  -11886.857763 +/- 20.516606   2673.399932 +/- 1580.638930   0.2249 
dmc-a128-e1536-cpu-DU24/NiO-fcc-S32-dmc  series 1  nan +/-      nan   nan +/-      nan      nan 
dmc-a128-e1536-cpu-DU32/NiO-fcc-S32-dmc  series 1  nan +/-      nan   nan +/-      nan      nan 
dmc-a128-e1536-cpu-DU48/NiO-fcc-S32-dmc  series 1  nan +/-      nan   nan +/-      nan      nan 
dmc-a128-e1536-cpu-DU64/NiO-fcc-S32-dmc  series 1  nan +/-      nan   nan +/-      nan      nan 
dmc-a128-e1536-cpu-DU80/NiO-fcc-S32-dmc  series 1  nan +/-      nan   nan +/-      nan      nan 
dmc-a128-e1536-cpu-DU96/NiO-fcc-S32-dmc  series 1  nan +/-      nan   nan +/-      nan      nan 
dmc-a128-e1536-cpu-DU128/NiO-fcc-S32-dmc  series 1  nan +/-      nan   nan +/-      nan      nan 
```

**Comment by ye-luo** [comment]
I was running c68fe6ada1c382c47339046adfe0fe3605b5d548. Please have a look. Thank you.

**Comment by atillack** [comment]
@ye-luo The short-NiO_a4_e48_pp-delayed_update-dmc_sdj-1-16 is failing for this reason:
`Fatal Error. Aborting at Tmove v1 and v3 have not been implemented on GPU.
241:   please contact the developers if you need this feature`

The other tests pass.

**Comment by prckent** [comment]
I made an issue for this separate fix

**Comment by atillack** [comment]
@ye-luo Thanks - I can replicate the NaNs in VMC w/o drift. Fix should be straightforward (jinx!) and is coming soon.

**Comment by atillack** [comment]
@ye-luo NaNs are fixed by using a pivot table for the inversions. I still need to run my full battery of tests but I am cautiously optimistic that this PR might be done :-)

**Comment by prckent** [comment]
@ye-luo Is this ready now?

**Comment by ye-luo** [comment]
I still got bad variance and sometimes NaN with VMC w/o drift. VMC with drift is working fine.
Since we need to do the code formatting. How about marking a safe guard and merge the code?
We can come back to fix the VMC w/o drift case.

**Comment by ye-luo** [comment]
```
                            LocalEnergy               Variance           ratio
dmc-a128-e1536-cpu-DU02-drift  -11865.775422 +/- 0.963954   358.098110 +/- 35.645283   0.0302 
dmc-a128-e1536-cpu-DU04-drift  -11868.841171 +/- 1.971000   380.407893 +/- 36.975904   0.0321 
dmc-a128-e1536-cpu-DU08-drift  -11865.781416 +/- 0.621689   486.364715 +/- 44.727172   0.0410 
dmc-a128-e1536-cpu-DU12-drift  -11868.032866 +/- 0.774748   371.279208 +/- 26.826092   0.0313 
dmc-a128-e1536-cpu-DU16-drift  -11868.036024 +/- 0.423531   376.381149 +/- 10.313711   0.0317 
dmc-a128-e1536-cpu-DU24-drift  -11868.038468 +/- 0.533695   343.597411 +/- 6.606281   0.0290 
dmc-a128-e1536-cpu-DU32-drift  -11867.028481 +/- 0.279506   390.566754 +/- 3.969441   0.0329 
dmc-a128-e1536-cpu-DU48-drift  -11867.638779 +/- 0.783277   403.974130 +/- 10.659912   0.0340 
dmc-a128-e1536-cpu-DU64-drift  -11866.831591 +/- 0.682510   371.089204 +/- 33.235639   0.0313 
dmc-a128-e1536-cpu-DU80-drift  -11867.336156 +/- 0.773123   441.516925 +/- 15.759158   0.0372 
dmc-a128-e1536-cpu-DU96-drift  -11867.977108 +/- 0.022990   388.242931 +/- 2.967783   0.0327 
dmc-a128-e1536-cpu-DU128-drift -11865.986793 +/- 1.053945   369.086747 +/- 34.109730   0.0311  

                            LocalEnergy               Variance           ratio
dmc-a128-e1536-cpu-DU02  -11862.660229 +/- 1.300478   413.323814 +/- 7.271194   0.0348 
dmc-a128-e1536-cpu-DU04  -11863.097625 +/- 1.235351   412.847536 +/- 27.300977   0.0348 
dmc-a128-e1536-cpu-DU08  -11864.847155 +/- 1.079036   402.869849 +/- 24.876893   0.0340 
dmc-a128-e1536-cpu-DU12  -11862.537028 +/- 0.523314   453.961613 +/- 8.637562   0.0383 
dmc-a128-e1536-cpu-DU16  -11863.134971 +/- 2.397491   410.008913 +/- 14.277702   0.0346 
dmc-a128-e1536-cpu-DU24  -11864.503837 +/- 2.379253   410.216990 +/- 51.665647   0.0346 
dmc-a128-e1536-cpu-DU32  -11862.718271 +/- 0.325593   433.520861 +/- 16.753439   0.0365 
dmc-a128-e1536-cpu-DU48  -11863.227816 +/- 1.732416   385.391294 +/- 34.742421   0.0325 
dmc-a128-e1536-cpu-DU64  -11862.919381 +/- 1.140548   378.458193 +/- 35.397991   0.0319 
dmc-a128-e1536-cpu-DU80  -11854.903590 +/- 2.578794   1378.816451 +/- 312.655708   0.1163 
dmc-a128-e1536-cpu-DU96  -11857.925186 +/- 0.142264   611.342749 +/- 18.367560   0.0516 
dmc-a128-e1536-cpu-DU128 -11857.841087 +/- 1.123030   916.823063 +/- 28.674093   0.0773 
```
@atillack Without drift, the DU 80/96/128 still carry high variance.

**Comment by atillack** [comment]
@ye-luo In the interest of getting this merged I would argue that beyond DU 32 there isn't any speed-up and at least up to DU 64 we're close in variance, so this may not be that much of a show stopper.

 I can switch to double precision (your calculation were mixed precision, right?) for those arrays like we do for mixed precision matrix inverses which should fix this. It will be ready tomorrow afternoon.

**Comment by ye-luo** [comment]
@atillack I'm OK merging this as it is. Regarding DU>64, I think it is not caused by numerical precision. I did check with full precision and DU>64 was bad as well. The with drift code works fine with mixed precision. My feeling is there might be a missing synchronization or something related to size 64.

Before merging the code, could you please add a safe-guard in the VMC w/o drift driver for size>64?

**Comment by atillack** [comment]
@ye-luo Thanks. Yes, I'll add a warning and cap DU at 64.

**Comment by ye-luo** [comment]
@atillack sadly, adding this stream synchronization doesn't help. I mean synchronization of threads within a kernel which has some launching parameter 64 (block size?).

**Comment by atillack** [comment]
@ye-luo Yeah, unfortunately there is no other somewhat suspicious location that stands out. I'll add the warning and cap at k = 64 and work on finding the fix later.

What makes me suspicious of numerical precision being the culprit is that with a Psi.recompute call after each step, everything seems fine (this is my NiO 128 atom data analyzed over the last 10 steps of 20 blocks of 1 step each):

                            LocalEnergy               Variance           ratio 
k = 1:
avg  series 1  -11782.962089 +/- 25.977604   473.905789 +/- 40.549188   0.0402 
k = 2:
avg  series 1  -11782.905483 +/- 25.761714   503.596322 +/- 47.569417   0.0427 
k = 4:
avg  series 1  -11783.675988 +/- 25.934824   498.788050 +/- 40.720720   0.0423 
k = 8:
avg  series 1  -11783.242652 +/- 25.829249   486.117116 +/- 44.136982   0.0413 
k = 16:
avg  series 1  -11783.116865 +/- 26.016353   498.418402 +/- 44.100940   0.0423 
k = 32:
avg  series 1  -11783.063903 +/- 25.766134   515.954829 +/- 42.950156   0.0438 
k = 64:
avg  series 1  -11782.884046 +/- 25.917708   486.847984 +/- 53.088808   0.0413 
k = 128:
avg  series 1  -11782.704666 +/- 25.630978   505.765987 +/- 34.084771   0.0429

**Comment by ye-luo** [comment]
Psi.recompute wipes out any bad stuff accumulated during the PbyP move. If the sampling goes wrong, the average value will goes wrong even if individual Psi is correct. Could you try to run the VMC winder(more nodes) and longer(more blocks) and also check k = 80?
Are you using 1 walker per GPU? Try 32.

**Comment by atillack** [comment]
I am running with 4 GPUs and 128 walkers/GPU.

**Comment by ye-luo** [comment]
I didn't understand why your error bar was around 25.
I'm expecting some value around sqrt(400/(4*128*20))= 0.197642354 Hartree.

**Comment by atillack** [comment]
Psi.recompute on GPU only updates the A inverse - this is why when AinvU (V'A^-1 * dU) and the lemma matrix where slightly inconsistent the SMW update (A^-1' = A^-1 - AinvU*Lemma^-1*(V'A^-1) started accumulating those errors leading to the observed NaNs.

I think there might still be some error accumulation going on that gets larger for bigger delay ranks. Part of the fix could be to go to higher precision for the SMW update (like we do for the full A^-1 update).

Here is the k = 80 run data (same as above, NiO 128 atoms, 4 GPUS, 128 walkers/GPU, 1 warmup step, 1 step per block, 20 blocks, last 10 are analyzed):
k = 80:
avg  series 1  -11782.911679 +/- 25.797730   479.054177 +/- 50.768536   0.0407

**Comment by atillack** [comment]
@ye-luo The error bar is likely that large as the VMC block I am running is the very first one and is not yet equilibrated. I get the same error bar that you get when I run the "official" input files.

**Comment by ye-luo** [comment]
The recomputing of A^-1 should not be the source. It is also applied on CPU mixed precision and I can safely go up-to k = 512. I start to worry about V100 since I was running on Summit. Were you running on SummitDev?

**Comment by atillack** [comment]
The current numbers are on SummitDev but I also get similar results on Summit. Btw, the k = 80 above was accidentally at k = 64 (the cap at 64 code works! and I forgot I already enabled it after lunch). Here is the k = 80 data:
                            LocalEnergy               Variance           ratio 
avg  series 1  -11784.798230 +/- 25.943351   548.457051 +/- 38.092702   0.0465

**Comment by atillack** [comment]
@ye-luo Here is what I get when i run with k=80 using the following VMC input block:

`<qmc method="vmc" move="pbyp" gpu="yes" kdelay="80">`
`    <estimator name="LocalEnergy" hdf5="no" />`
`    <parameter name="walkers">128</parameter>`
`    <parameter name="stepsbetweensamples">    1 </parameter>`
`    <parameter name="warmupSteps">            5 </parameter>`
`    <parameter name="substeps">               5 </parameter>`
`    <parameter name="steps">                  2 </parameter>`
`    <parameter name="blocks">                 5 </parameter>`
`    <parameter name="timestep">             1.0 </parameter>`
`    <parameter name="usedrift">              no </parameter>`
`</qmc>`

                            LocalEnergy               Variance           ratio 
NiO-fcc-S32-vmc  series 1  -11865.977789 +/- 0.700271   408.558316 +/- 4.459451   0.0344 

What is your VMC block?

**Comment by ye-luo** [comment]
@atillack Your VMC block seems fine. I don't put the kdelay via VMC block. I did `<slaterdeterminant delay_rank="80">` The error bar in your last rely seems reasonable. Anyway this is not the real issue.
Your k = 80 results confuse me. Are they actually all capped as k = 64?
What is the real k = 80 result? Is it problematic on SummitDev?

**Comment by atillack** [comment]
@ye-luo I have working results up to k=128 bit I also run into the diverging variance at k=256. From my experience, recompute during the warmup steps helps which does make me believe it is numerical instability when more than a handful of steps go by with continuous usage of SMW updates rather than full ones.

**Comment by ye-luo** [comment]
@atillack I tried Summitdev and I got correct numbers for k=80,96,128. So I believe the issue I'm encountering is related to Summit.
@prckent I'm afraid some kernels written for pre-Volta architectures may be no more safe on Volta. I have not gotten a chance to try Titan yet.
Unless there is any other concern. I will approve and merge the code tomorrow. CI machine is under maintenance today.

**Comment by atillack** [comment]
@ye-luo @prckent I am running on Titan right now.

**Comment by atillack** [comment]
@ye-luo @prckent It's fixed and working now!

Thanks @ye-luo for the idea of looking for CUDA threading issues, in the kernel to finish calculation of the lemma and ainvu matrices there was the possibility of changing data underneath another thread trying to use that data...

Here is my current SummitDev @ k=256 run:
k = 256:
NiO-fcc-S32-vmc  series 1  -11865.801500 +/- 0.545776   426.167319 +/- 23.228991   0.0359

Rerunning the other tests...

**Comment by atillack** [comment]
@ye-luo @prckent Everything is working. Here is my current VMC w/o drift series on SummitDev:

                            LocalEnergy               Variance           ratio 
k = 1:
NiO-fcc-S32-vmc  series 1  -11865.578301 +/- 0.851429   404.349249 +/- 11.021452   0.0341 
k = 2:
NiO-fcc-S32-vmc  series 1  -11865.349947 +/- 0.916794   396.919920 +/- 7.895504   0.0335 
k = 4:
NiO-fcc-S32-vmc  series 1  -11866.000019 +/- 0.772399   424.014092 +/- 22.672240   0.0357 
k = 8:
NiO-fcc-S32-vmc  series 1  -11865.711592 +/- 0.675608   403.633995 +/- 5.521675   0.0340 
k = 16:
NiO-fcc-S32-vmc  series 1  -11865.642270 +/- 0.453154   401.845648 +/- 11.238411   0.0339 
k = 32:
NiO-fcc-S32-vmc  series 1  -11865.794974 +/- 0.935828   409.392814 +/- 12.941130   0.0345 
k = 64:
NiO-fcc-S32-vmc  series 1  -11865.310495 +/- 0.519327   403.773517 +/- 11.573975   0.0340 
k = 128:
NiO-fcc-S32-vmc  series 1  -11865.702223 +/- 0.690361   421.433592 +/- 6.115831   0.0355 
k = 256:
NiO-fcc-S32-vmc  series 1  -11865.801500 +/- 0.545776   426.167319 +/- 23.228991   0.0359

**Comment by ye-luo** [comment]
I verified that runs with up-to k = 256 are correct on summit.

**Pull request merged**